File: ios/LoqaAudioDspModule.swift
import ExpoModulesCore

public class LoqaAudioDspModule: Module {
  // Module definition for Expo Modules API
  public func definition() -> ModuleDefinition {
    // Module name that JavaScript will use to require this module
    Name("LoqaAudioDsp")

    // MARK: - computeFFT (Implemented in Story 2.2)
    // Calls Rust FFT via RustBridge.computeFFTWrapper()
    AsyncFunction("computeFFT") { (buffer: [Float], options: [String: Any], promise: Promise) in
      do {
        // Extract options with defaults
        let fftSize = options["fftSize"] as? Int ?? buffer.count
        let windowTypeString = options["windowType"] as? String ?? "hanning"
        let sampleRate = options["sampleRate"] as? Double ?? 44100.0

        // Validate inputs
        guard !buffer.isEmpty else {
          promise.reject("VALIDATION_ERROR", "Buffer cannot be empty", nil)
          return
        }

        guard fftSize > 0 && (fftSize & (fftSize - 1)) == 0 else {
          promise.reject("VALIDATION_ERROR", "FFT size must be a power of 2", nil)
          return
        }

        guard fftSize >= 256 && fftSize <= 8192 else {
          promise.reject("VALIDATION_ERROR", "FFT size must be between 256 and 8192", nil)
          return
        }

        // Map windowType string to integer (none=0, hanning=1, hamming=2, blackman=3)
        let windowType: Int32
        switch windowTypeString.lowercased() {
        case "none":
          windowType = 0
        case "hanning":
          windowType = 1
        case "hamming":
          windowType = 2
        case "blackman":
          windowType = 3
        default:
          promise.reject("VALIDATION_ERROR", "Invalid window type. Must be one of: none, hanning, hamming, blackman", nil)
          return
        }

        // Call Rust FFT function via wrapper
        let magnitude = try computeFFTWrapper(buffer: buffer, fftSize: fftSize, windowType: windowType)

        // Build frequency array: freq[i] = (sampleRate / fftSize) * i
        let frequencies = (0..<magnitude.count).map { Float(sampleRate / Double(fftSize) * Double($0)) }

        // Return result dictionary
        let result: [String: Any] = [
          "magnitude": magnitude,
          "frequencies": frequencies
        ]

        promise.resolve(result)
      } catch let error as RustFFIError {
        // Handle Rust FFI errors with specific error codes
        switch error {
        case .invalidInput(let message):
          promise.reject("VALIDATION_ERROR", message, nil)
        case .computationFailed(let message):
          promise.reject("FFT_ERROR", message, nil)
        case .memoryAllocationFailed:
          promise.reject("FFT_ERROR", "Memory allocation failed in Rust FFT computation", nil)
        }
      } catch {
        // Handle unexpected errors
        promise.reject("FFT_ERROR", error.localizedDescription, error)
      }
    }

    // MARK: - detectPitch (Implemented in Story 3.3)
    // Calls Rust YIN pitch detection via RustBridge.detectPitchWrapper()
    AsyncFunction("detectPitch") { (buffer: [Float], sampleRate: Int, options: [String: Any], promise: Promise) in
      do {
        // Validate inputs (basic validation - detailed validation in RustBridge)
        guard !buffer.isEmpty else {
          promise.reject("VALIDATION_ERROR", "Buffer cannot be empty")
          return
        }

        guard sampleRate >= 8000 && sampleRate <= 48000 else {
          promise.reject("VALIDATION_ERROR", "Sample rate must be between 8000 and 48000 Hz")
          return
        }

        // Call Rust pitch detection via wrapper
        let (frequency, confidence, isVoiced) = try detectPitchWrapper(
          buffer: buffer,
          sampleRate: sampleRate
        )

        // Build result dictionary matching PitchResult type
        // frequency is Optional<Float>, convert to NSNull if nil
        let result: [String: Any] = [
          "frequency": frequency ?? NSNull(),
          "confidence": confidence,
          "isVoiced": isVoiced
        ]

        promise.resolve(result)
      } catch let error as RustFFIError {
        // Handle Rust FFI errors with specific error codes
        switch error {
        case .invalidInput(let message):
          promise.reject("VALIDATION_ERROR", message)
        case .computationFailed(let message):
          promise.reject("PITCH_ERROR", message)
        case .memoryAllocationFailed:
          promise.reject("PITCH_ERROR", "Memory allocation failed in Rust pitch detection")
        }
      } catch {
        // Handle unexpected errors
        promise.reject("PITCH_ERROR", error.localizedDescription, error)
      }
    }

    // MARK: - extractFormants (Implemented in Story 3.3)
    // Calls Rust LPC formant extraction via RustBridge.extractFormantsWrapper()
    AsyncFunction("extractFormants") { (buffer: [Float], sampleRate: Int, options: [String: Any], promise: Promise) in
      do {
        // Extract optional LPC order from options
        let lpcOrder = options["lpcOrder"] as? Int

        // Validate inputs (basic validation - detailed validation in RustBridge)
        guard !buffer.isEmpty else {
          promise.reject("VALIDATION_ERROR", "Buffer cannot be empty")
          return
        }

        guard sampleRate >= 8000 && sampleRate <= 48000 else {
          promise.reject("VALIDATION_ERROR", "Sample rate must be between 8000 and 48000 Hz")
          return
        }

        // Call Rust formant extraction via wrapper
        let (f1, f2, f3, bandwidths) = try extractFormantsWrapper(
          buffer: buffer,
          sampleRate: sampleRate,
          lpcOrder: lpcOrder
        )

        // Build result dictionary matching FormantsResult type
        let result: [String: Any] = [
          "f1": f1,
          "f2": f2,
          "f3": f3,
          "bandwidths": [
            "f1": bandwidths.0,
            "f2": bandwidths.1,
            "f3": bandwidths.2
          ]
        ]

        promise.resolve(result)
      } catch let error as RustFFIError {
        // Handle Rust FFI errors with specific error codes
        switch error {
        case .invalidInput(let message):
          promise.reject("VALIDATION_ERROR", message)
        case .computationFailed(let message):
          promise.reject("FORMANTS_ERROR", message)
        case .memoryAllocationFailed:
          promise.reject("FORMANTS_ERROR", "Memory allocation failed in Rust formant extraction")
        }
      } catch {
        // Handle unexpected errors
        promise.reject("FORMANTS_ERROR", error.localizedDescription, error)
      }
    }

    // MARK: - analyzeSpectrum (Implemented in Story 4.2)
    // Calls Rust spectral analysis via RustBridge.analyzeSpectrumWrapper()
    AsyncFunction("analyzeSpectrum") { (buffer: [Float], sampleRate: Int, options: [String: Any], promise: Promise) in
      do {
        // Validate inputs (basic validation - detailed validation in RustBridge)
        guard !buffer.isEmpty else {
          promise.reject("VALIDATION_ERROR", "Buffer cannot be empty")
          return
        }

        guard sampleRate >= 8000 && sampleRate <= 48000 else {
          promise.reject("VALIDATION_ERROR", "Sample rate must be between 8000 and 48000 Hz")
          return
        }

        // Call Rust spectral analysis via wrapper
        let (centroid, rolloff, tilt) = try analyzeSpectrumWrapper(
          buffer: buffer,
          sampleRate: sampleRate
        )

        // Build result dictionary matching SpectrumResult type
        let result: [String: Any] = [
          "centroid": centroid,
          "rolloff": rolloff,
          "tilt": tilt
        ]

        promise.resolve(result)
      } catch let error as RustFFIError {
        // Handle Rust FFI errors with specific error codes
        switch error {
        case .invalidInput(let message):
          promise.reject("VALIDATION_ERROR", message)
        case .computationFailed(let message):
          promise.reject("SPECTRUM_ERROR", message)
        case .memoryAllocationFailed:
          promise.reject("SPECTRUM_ERROR", "Memory allocation failed in Rust spectrum analysis")
        }
      } catch {
        // Handle unexpected errors
        promise.reject("SPECTRUM_ERROR", error.localizedDescription, error)
      }
    }
  }
}

File: android/src/main/java/com/loqalabs/loquaaudiodsp/LoqaAudioDspModule.kt
package com.loqalabs.loquaaudiodsp

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.loqalabs.loquaaudiodsp.RustJNI.RustBridge

/**
 * LoqaAudioDspModule provides Expo Module API for accessing Rust DSP functions.
 *
 * This module exposes async functions for:
 * - computeFFT: Fast Fourier Transform analysis
 * - detectPitch: YIN pitch detection algorithm
 * - extractFormants: LPC formant extraction
 * - analyzeSpectrum: Spectral feature analysis
 *
 * All functions run on background threads automatically via Expo's AsyncFunction.
 * Results are returned via Promises for async/await support in JavaScript/TypeScript.
 *
 * Implementation Notes:
 * - Story 1.4: Placeholder async function stubs (this story)
 * - Story 2.3: Real FFT implementation
 * - Story 3.3: Real pitch and formant implementations
 * - Story 4.2: Real spectrum analysis implementation
 */
class LoqaAudioDspModule : Module() {
  // Module definition for Expo Modules API
  override fun definition() = ModuleDefinition {
    // Module name that JavaScript will use to require this module
    Name("LoqaAudioDsp")

    // ============================================================================
    // Async Function: computeFFT
    // ============================================================================

    /**
     * Computes Fast Fourier Transform (FFT) on audio buffer.
     *
     * Implementation completed in Story 2.3.
     * Expo automatically runs this on a background thread.
     *
     * @param buffer Audio samples as FloatArray
     * @param options Map with optional keys: "fftSize" (Int), "windowType" (String), "sampleRate" (Double)
     * @return Map with keys: "magnitude" (FloatArray), "frequencies" (FloatArray)
     * @throws Exception with error code "VALIDATION_ERROR" or "FFT_ERROR"
     */
    AsyncFunction("computeFFT") { buffer: FloatArray, options: Map<String, Any?> ->
      // Extract options with defaults
      val fftSize = (options["fftSize"] as? Int) ?: buffer.size
      val windowTypeString = (options["windowType"] as? String) ?: "hanning"
      val sampleRate = (options["sampleRate"] as? Double) ?: 44100.0

      // Validate buffer is not empty (AC2)
      if (buffer.isEmpty()) {
        throw Exception("VALIDATION_ERROR: Buffer cannot be empty")
      }

      // Validate fftSize is power of 2 (AC2)
      if (fftSize <= 0 || (fftSize and (fftSize - 1)) != 0) {
        throw Exception("VALIDATION_ERROR: FFT size must be a power of 2, got $fftSize")
      }

      // Validate fftSize is in range 256-8192 (AC2)
      if (fftSize < 256 || fftSize > 8192) {
        throw Exception("VALIDATION_ERROR: FFT size must be between 256 and 8192, got $fftSize")
      }

      // Map windowType string to integer (none=0, hanning=1, hamming=2, blackman=3)
      val windowType: Int = when (windowTypeString.lowercase()) {
        "none" -> 0
        "hanning" -> 1
        "hamming" -> 2
        "blackman" -> 3
        else -> throw Exception("VALIDATION_ERROR: Invalid window type '$windowTypeString'. Must be one of: none, hanning, hamming, blackman")
      }

      try {
        // Call Rust FFT function via JNI (AC3)
        // JNI handles FloatArray marshalling automatically (AC4)
        val magnitude = RustBridge.computeFFT(buffer, fftSize, windowType)

        // Build frequencies array: freq[i] = (sampleRate / fftSize) * i (AC5)
        val frequencies = FloatArray(magnitude.size) { i ->
          (sampleRate / fftSize * i).toFloat()
        }

        // Return result map (AC5)
        mapOf(
          "magnitude" to magnitude,
          "frequencies" to frequencies
        )
      } catch (e: RuntimeException) {
        // Catch JNI/Rust errors and reject with FFT_ERROR code (AC6)
        throw Exception("FFT_ERROR: ${e.message}", e)
      } catch (e: Exception) {
        // Catch any other unexpected errors (AC6)
        throw Exception("FFT_ERROR: Unexpected error during FFT computation - ${e.message}", e)
      }
    }

    // ============================================================================
    // Async Function: detectPitch
    // ============================================================================

    /**
     * Detects pitch using YIN algorithm.
     *
     * Placeholder implementation - will be completed in Story 3.3.
     * Expo automatically runs this on a background thread.
     *
     * @param buffer Audio samples as FloatArray
     * @param sampleRate Sample rate in Hz (Int)
     * @param options Map with optional keys: "minFrequency" (Float), "maxFrequency" (Float)
     * @return Map with keys: "frequency" (Float), "confidence" (Float), "isVoiced" (Boolean)
     * @throws Exception with error code "PITCH_ERROR"
     */
    AsyncFunction("detectPitch") { buffer: FloatArray, sampleRate: Int, options: Map<String, Any?> ->
      try {
        // Placeholder: This will call RustBridge.detectPitch in Story 3.3
        throw UnsupportedOperationException(
          "detectPitch not yet implemented. Will be completed in Story 3.3."
        )
      } catch (e: Exception) {
        throw Exception("PITCH_ERROR: ${e.message}", e)
      }
    }

    // ============================================================================
    // Async Function: extractFormants
    // ============================================================================

    /**
     * Extracts formant frequencies (F1, F2, F3) using LPC analysis.
     *
     * Placeholder implementation - will be completed in Story 3.3.
     * Expo automatically runs this on a background thread.
     *
     * @param buffer Audio samples as FloatArray
     * @param sampleRate Sample rate in Hz (Int)
     * @param options Map with optional keys: "lpcOrder" (Int)
     * @return Map with keys: "f1" (Float), "f2" (Float), "f3" (Float), "bandwidths" (Map)
     * @throws Exception with error code "FORMANTS_ERROR"
     */
    AsyncFunction("extractFormants") { buffer: FloatArray, sampleRate: Int, options: Map<String, Any?> ->
      try {
        // Placeholder: This will call RustBridge.extractFormants in Story 3.3
        throw UnsupportedOperationException(
          "extractFormants not yet implemented. Will be completed in Story 3.3."
        )
      } catch (e: Exception) {
        throw Exception("FORMANTS_ERROR: ${e.message}", e)
      }
    }

    // ============================================================================
    // Async Function: analyzeSpectrum
    // ============================================================================

    /**
     * Analyzes spectral features (centroid, rolloff, tilt).
     *
     * Placeholder implementation - will be completed in Story 4.2.
     * Expo automatically runs this on a background thread.
     *
     * @param buffer Audio samples as FloatArray
     * @param sampleRate Sample rate in Hz (Int)
     * @param options Map with optional configuration
     * @return Map with keys: "centroid" (Float), "rolloff" (Float), "tilt" (Float)
     * @throws Exception with error code "SPECTRUM_ERROR"
     */
    AsyncFunction("analyzeSpectrum") { buffer: FloatArray, sampleRate: Int, options: Map<String, Any?> ->
      try {
        // Placeholder: This will call RustBridge.analyzeSpectrum in Story 4.2
        throw UnsupportedOperationException(
          "analyzeSpectrum not yet implemented. Will be completed in Story 4.2."
        )
      } catch (e: Exception) {
        throw Exception("SPECTRUM_ERROR: ${e.message}", e)
      }
    }
  }
}
